cloud/docker/docker_swarm_service.py

```diff
@@ -0,0 +1,615 @@
+      self.mode="replicated"
+      self.mounts=[]
+      self.constraints=[]
+      self.networks=[]
+      self.publish=[]
+      self.replicas=1
+      self.service_id=False
+      self.service_version=False
+      self.restart_policy = None
+      self.restart_policy_attempts = None
+      self.restart_policy_delay = None
+      self.restart_policy_window = None
+
+    def __str__(self):
+      return str({
+        'mode': self.mode,
+        'env': self.env,
+        'endpoint_mode': self.endpoint_mode,
+        'mounts': self.mounts,
+        'networks': self.networks,
+        'replicas': self.replicas
+      })
+    def generate_docker_py_service_description(self, name, docker_networks):
+        cspec={
+          'Image': self.image
+        }
+        cspec['Mounts']=[]
+        for mount_config in self.mounts:
+          cspec['Mounts'].append({
+            'Target': mount_config['target'],
+            'Source': mount_config['source'],
+            'Type': mount_config['type']
+          })
+        cspec['Args']=self.args
+        cspec['Env']=self.env
+        cspec['Labels']=self.container_labels
+        restart_policy=docker.types.RestartPolicy(
+          condition     = self.restart_policy,
+          delay         = self.restart_policy_delay,
+          max_attempts  = self.restart_policy_attempts,
+          window        = self.restart_policy_window
+        )
+        resources={
+          'Limits': {
+            'NanoCPUs': int(self.limit_cpu*1000000000),
+            'MemoryBytes': self.limit_memory*1024*1024
+          },
+          'Reservations': {
+            'NanoCPUs': int(self.limit_cpu*1000000000),
+            'MemoryBytes': self.limit_memory*1024*1024
+          }
+        }
+        task_template=docker.types.TaskTemplate(
+          container_spec  = cspec,
+          restart_policy  = restart_policy,
+          placement       = self.constraints,
+          resources       = resources
+        )
+        mode = { 'Replicated': {'Replicas': self.replicas} }
+        if self.mode=='global':
+          mode = { 'Global': {} }
+
+        networks=[]
+        for network_name in self.networks:
+          network_id = None
+          try:
+            network_id = filter( lambda n: n['name']==network_name, docker_networks )[0]['id']
+          except:
+            pass
+          if network_id:
+            networks.append (
+              { 'Target': network_id }
+            )
+          else:
+            raise Exception("no docker networks named: %s" % network_name)
+
+        endpoint_spec= {'Mode': self.endpoint_mode }
+        endpoint_spec['Ports'] = []
+        for port in self.publish:
+          endpoint_spec['Ports'].append({
+            'Protocol': port['protocol'],
+            'PublishedPort': port['published_port'],
+            'TargetPort': port['target_port']
+          })
+        return task_template, networks, endpoint_spec, mode, self.labels
+
+  def __init__(self, socket_path=None):
```

Comment:
I'm nearly done reingeneering the module for using the classes defined in docker_common.py (as suggested by @mscherer ), thus making those shared parameters available in this module.